packages/next/src/server/app-render/action-handler.ts

```diff
@@ -216,22 +216,38 @@ async function createRedirectRenderResult(
       if (
         headResponse.headers.get('content-type') === RSC_CONTENT_TYPE_HEADER
       ) {
-        const response = await fetch(fetchUrl, {
+        const redirectResponse = await fetch(fetchUrl, {
           method: 'GET',
           headers: forwardedHeaders,
           next: {
             // @ts-ignore
             internal: 1,
           },
         })
+        //copy cookies from the redirect response to respose we're sending before copying other headers
+        //this is needed only in case of custom server created by developer
+        //cause nextjs app dir does not allow to set cookies from components
+        const redirectResponseCookies = new ResponseCookies(
+          redirectResponse.headers
+        )
+        const responseCookies = new ResponseCookies(
+          fromNodeOutgoingHttpHeaders(res.getHeaders())
+        )
```

Comment:
It doesn't seem like `responseCookies` is getting used?

Reply:
No its getting used on line 238 and 240. And `res` is the final response that browser receives as response to server action + redirect invocation hence have to carry those cookies from redirectResponse to response.

Reply:
I saw that it's mutated (`.delete` and `.set`) on these 2 lines, but `responseCookies` itself is never get used/returned?

Reply:
Understood. we used to copy cookie header as a whole before as per [this line](https://github.com/vercel/next.js/pull/61633/files#diff-c809d50461027cdba7c092e564818b1172133d337abc5c513f829c94c8483dc6L230) but it overrides cookie header fully instead of merging cookies. Hence I made this change to merge cookies. 

While your question is valid that responseCookies does not seem to be used but those are cookies from `res`. and this `res` is getting in turn returned to browser via [this line](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/app-render/app-render.tsx#L1318) eventually dropping the merged cookies on browser.